flytekit/core/rust/fs.py

```diff
@@ -0,0 +1,26 @@
+import asyncio
+from s3fs import S3FileSystem
+from fsspec.callbacks import _DEFAULT_CALLBACK
+import rustfs
+
+class RustS3FileSystem(S3FileSystem):
```

Comment:
should we re-implement the entire s3Filesystem? it seems s3Filesystem is inefficient? 
Isn't this multi-part effectively serialized because of the `await` https://github.com/fsspec/s3fs/blob/a14c31a382c50996aff37107245768a13efef0cb/s3fs/core.py#L1179?

Reply:
ohh i see, you have overriden __put_file. 
Can you test it against a pure python implementation that does not `await` every chunk, but awaits the futures after all chunks of enqueued?

Reply:
NM i think I am wrong, cc @EngHabu I see that s3fs uses the aiobotocore, and I think upload_part in aiobotocore returns a stream once established

Reply:
I think s3fs reads the chunks one by one because it await for every read. https://github.com/fsspec/s3fs/blob/a14c31a382c50996aff37107245768a13efef0cb/s3fs/core.py#L1234,

Reply:
But one thing we can test is what you mentioned, using pure python implementation and awaits the futures after all chunks are enqueued.

Reply:
FYI, ttps://github.com/fsspec/s3fs/issues/796

Reply:
Hi @kumare3, I have finished the pure python implementation. Please check the latest comment.